File: mtr_app/generators/corridor_calc.py
"""
Geospatial Corridor and Geodesy Calculation Module for Military Training Routes.
Computes bearings, distances, perpendicular corridor boundary polygons, and coordinate conversions.
"""

import math
from typing import List, Tuple, Dict, Any

EARTH_RADIUS_NM = 3440.065  # WGS-84 mean radius in Nautical Miles


def calculate_bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculates initial great-circle bearing from (lat1, lon1) to (lat2, lon2) in degrees (0-360).
    """
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_lambda = math.radians(lon2 - lon1)

    y = math.sin(delta_lambda) * math.cos(phi2)
    x = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(delta_lambda)
    theta = math.atan2(y, x)
    bearing = (math.degrees(theta) + 360.0) % 360.0
    return round(bearing, 2)


def calculate_distance_nm(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculates great-circle distance between two points in Nautical Miles using Haversine formula.
    """
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)

    a = math.sin(delta_phi / 2.0) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2.0) ** 2
    c = 2.0 * math.atan2(math.sqrt(a), math.sqrt(1.0 - a))
    distance = EARTH_RADIUS_NM * c
    return round(distance, 2)


def destination_point(lat: float, lon: float, distance_nm: float, bearing_deg: float) -> Tuple[float, float]:
    """
    Calculates destination point given start point, distance (NM), and bearing (degrees).
    Returns (lat, lon) in decimal degrees.
    """
    delta = distance_nm / EARTH_RADIUS_NM
    theta = math.radians(bearing_deg)
    phi1 = math.radians(lat)
    lambda1 = math.radians(lon)

    phi2 = math.asin(
        math.sin(phi1) * math.cos(delta) +
        math.cos(phi1) * math.sin(delta) * math.cos(theta)
    )
    lambda2 = lambda1 + math.atan2(
        math.sin(theta) * math.sin(delta) * math.cos(phi1),
        math.cos(delta) - math.sin(phi1) * math.sin(phi2)
    )

    # Normalize longitude to -180..+180
    lon_deg = (math.degrees(lambda2) + 540.0) % 360.0 - 180.0
    lat_deg = math.degrees(phi2)
    return (round(lat_deg, 6), round(lon_deg, 6))


def to_arinc_dms(lat: float, lon: float) -> Tuple[str, str]:
    """
    Converts decimal lat/lon into ARINC 424 standard coordinate representations:
    Latitude: DDMMSSss[N|S] e.g. '34542016N' (34 deg, 54 min, 20.16 sec N)
    Longitude: DDDMMSSss[E|W] e.g. '117525556W' (117 deg, 52 min, 55.56 sec W)
    """
    # Latitude
    lat_hem = "N" if lat >= 0 else "S"
    lat_abs = abs(lat)
    lat_deg = int(lat_abs)
    lat_rem = (lat_abs - lat_deg) * 60.0
    lat_min = int(lat_rem)
    lat_sec = (lat_rem - lat_min) * 60.0
    lat_sec_int = int(round(lat_sec * 100))
    if lat_sec_int >= 6000:
        lat_sec_int = 0
        lat_min += 1
    lat_str = f"{lat_deg:02d}{lat_min:02d}{lat_sec_int:04d}{lat_hem}"

    # Longitude
    lon_hem = "E" if lon >= 0 else "W"
    lon_abs = abs(lon)
    lon_deg = int(lon_abs)
    lon_rem = (lon_abs - lon_deg) * 60.0
    lon_min = int(lon_rem)
    lon_sec = (lon_rem - lon_min) * 60.0
    lon_sec_int = int(round(lon_sec * 100))
    if lon_sec_int >= 6000:
        lon_sec_int = 0
        lon_min += 1
    lon_str = f"{lon_deg:03d}{lon_min:02d}{lon_sec_int:04d}{lon_hem}"

    return (lat_str, lon_str)


def to_human_dms(lat: float, lon: float) -> Tuple[str, str]:
    """
    Converts decimal lat/lon into human-readable DMS string:
    e.g. 34° 54' 20.16\" N, 117° 52' 55.56\" W
    """
    lat_hem = "N" if lat >= 0 else "S"
    lat_abs = abs(lat)
    lat_deg = int(lat_abs)
    lat_rem = (lat_abs - lat_deg) * 60.0
    lat_min = int(lat_rem)
    lat_sec = round((lat_rem - lat_min) * 60.0, 2)
    lat_str = f"{lat_deg:02d}° {lat_min:02d}' {lat_sec:05.2f}\" {lat_hem}"

    lon_hem = "E" if lon >= 0 else "W"
    lon_abs = abs(lon)
    lon_deg = int(lon_abs)
    lon_rem = (lon_abs - lon_deg) * 60.0
    lon_min = int(lon_rem)
    lon_sec = round((lon_rem - lon_min) * 60.0, 2)
    lon_str = f"{lon_deg:03d}° {lon_min:02d}' {lon_sec:05.2f}\" {lon_hem}"

    return (lat_str, lon_str)


def generate_corridor_polygon(segments: List[Dict[str, Any]]) -> List[List[float]]:
    """
    Generates a closed polygon ring [ [lon, lat], ... ] representing the corridor
    ribbon around the route centerline based on left/right widths in NM.
    """
    if not segments:
        return []

    # Build sequence of ordered waypoints and segment widths
    # For N segments, we have N+1 waypoints (or N legs)
    left_points = []
    right_points = []

    for i, seg in enumerate(segments):
        lat1 = seg.get("latitude_dec")
        lon1 = seg.get("longitude_dec")
        lat2 = seg.get("next_lat_dec")
        lon2 = seg.get("next_lon_dec")

        # If next point is not in segment row, check next segment or use course
        if lat2 is None or lon2 is None:
            if i + 1 < len(segments):
                lat2 = segments[i + 1].get("latitude_dec")
                lon2 = segments[i + 1].get("longitude_dec")
            else:
                # Last segment with no next coords, continue with previous bearing
                break

        width_left = seg.get("width_left_nm", 5.0) or 5.0
        width_right = seg.get("width_right_nm", 5.0) or 5.0

        course = calculate_bearing(lat1, lon1, lat2, lon2)
        bearing_left = (course - 90.0) % 360.0
        bearing_right = (course + 90.0) % 360.0

        # Start of segment offsets
        start_left = destination_point(lat1, lon1, width_left, bearing_left)
        start_right = destination_point(lat1, lon1, width_right, bearing_right)

        # End of segment offsets
        end_left = destination_point(lat2, lon2, width_left, bearing_left)
        end_right = destination_point(lat2, lon2, width_right, bearing_right)

        left_points.append(start_left)
        if i == len(segments) - 1 or (i + 1 < len(segments) and segments[i + 1].get("next_lat_dec") is None):
            left_points.append(end_left)

        right_points.append(start_right)
        if i == len(segments) - 1 or (i + 1 < len(segments) and segments[i + 1].get("next_lat_dec") is None):
            right_points.append(end_right)

    if not left_points or not right_points:
        return []

    # Form closed polygon: left boundary forward, then right boundary in reverse, closing at start
    polygon_coords = []
    for lat, lon in left_points:
        polygon_coords.append([lon, lat])

    for lat, lon in reversed(right_points):
        polygon_coords.append([lon, lat])

    # Close ring
    if polygon_coords:
        polygon_coords.append(polygon_coords[0])

    return polygon_coords

File: src/generators/corridorCalc.ts
import { MTRSegment } from "../types.js";

export const EARTH_RADIUS_NM = 3440.065; // WGS-84 mean radius in Nautical Miles

export function calculateBearing(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;

  const y = Math.sin(deltaLambda) * Math.cos(phi2);
  const x = Math.cos(phi1) * Math.sin(phi2) - Math.sin(phi1) * Math.cos(phi2) * Math.cos(deltaLambda);
  const theta = Math.atan2(y, x);
  const bearing = ((theta * 180) / Math.PI + 360.0) % 360.0;
  return Math.round(bearing * 100) / 100;
}

export function calculateDistanceNm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const deltaPhi = ((lat2 - lat1) * Math.PI) / 180;
  const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(deltaPhi / 2.0) ** 2 +
    Math.cos(phi1) * Math.cos(phi2) * Math.sin(deltaLambda / 2.0) ** 2;
  const c = 2.0 * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));
  const distance = EARTH_RADIUS_NM * c;
  return Math.round(distance * 100) / 100;
}

export function destinationPoint(
  lat: number,
  lon: number,
  distanceNm: number,
  bearingDeg: number
): [number, number] {
  const delta = distanceNm / EARTH_RADIUS_NM;
  const theta = (bearingDeg * Math.PI) / 180;
  const phi1 = (lat * Math.PI) / 180;
  const lambda1 = (lon * Math.PI) / 180;

  const phi2 = Math.asin(
    Math.sin(phi1) * Math.cos(delta) + Math.cos(phi1) * Math.sin(delta) * Math.cos(theta)
  );
  const lambda2 =
    lambda1 +
    Math.atan2(
      Math.sin(theta) * Math.sin(delta) * Math.cos(phi1),
      Math.cos(delta) - Math.sin(phi1) * Math.sin(phi2)
    );

  const lonDeg = (((lambda2 * 180) / Math.PI + 540.0) % 360.0) - 180.0;
  const latDeg = (phi2 * 180) / Math.PI;
  return [Math.round(latDeg * 1e6) / 1e6, Math.round(lonDeg * 1e6) / 1e6];
}

export function toArincDms(lat: number, lon: number): [string, string] {
  const latHem = lat >= 0 ? "N" : "S";
  const latAbs = Math.abs(lat);
  let latDeg = Math.floor(latAbs);
  const latRem = (latAbs - latDeg) * 60.0;
  let latMin = Math.floor(latRem);
  const latSec = (latRem - latMin) * 60.0;
  let latSecInt = Math.round(latSec * 100);
  if (latSecInt >= 6000) {
    latSecInt = 0;
    latMin += 1;
  }
  if (latMin >= 60) {
    latMin = 0;
    latDeg += 1;
  }
  const latStr = `${String(latDeg).padStart(2, "0")}${String(latMin).padStart(2, "0")}${String(latSecInt).padStart(4, "0")}${latHem}`;

  const lonHem = lon >= 0 ? "E" : "W";
  const lonAbs = Math.abs(lon);
  let lonDeg = Math.floor(lonAbs);
  const lonRem = (lonAbs - lonDeg) * 60.0;
  let lonMin = Math.floor(lonRem);
  const lonSec = (lonRem - lonMin) * 60.0;
  let lonSecInt = Math.round(lonSec * 100);
  if (lonSecInt >= 6000) {
    lonSecInt = 0;
    lonMin += 1;
  }
  if (lonMin >= 60) {
    lonMin = 0;
    lonDeg += 1;
  }
  const lonStr = `${String(lonDeg).padStart(3, "0")}${String(lonMin).padStart(2, "0")}${String(lonSecInt).padStart(4, "0")}${lonHem}`;

  return [latStr, lonStr];
}

export function toHumanDms(lat: number, lon: number): [string, string] {
  const latHem = lat >= 0 ? "N" : "S";
  const latAbs = Math.abs(lat);
  let latDeg = Math.floor(latAbs);
  const latRem = (latAbs - latDeg) * 60.0;
  let latMin = Math.floor(latRem);
  let latSec = Math.round((latRem - latMin) * 60.0 * 100) / 100;
  if (latSec >= 60) {
    latSec = 0;
    latMin += 1;
  }
  if (latMin >= 60) {
    latMin = 0;
    latDeg += 1;
  }
  const latSecStr = latSec < 10 ? `0${latSec.toFixed(2)}` : latSec.toFixed(2);
  const latStr = `${String(latDeg).padStart(2, "0")}° ${String(latMin).padStart(2, "0")}' ${latSecStr}" ${latHem}`;

  const lonHem = lon >= 0 ? "E" : "W";
  const lonAbs = Math.abs(lon);
  let lonDeg = Math.floor(lonAbs);
  const lonRem = (lonAbs - lonDeg) * 60.0;
  let lonMin = Math.floor(lonRem);
  let lonSec = Math.round((lonRem - lonMin) * 60.0 * 100) / 100;
  if (lonSec >= 60) {
    lonSec = 0;
    lonMin += 1;
  }
  if (lonMin >= 60) {
    lonMin = 0;
    lonDeg += 1;
  }
  const lonSecStr = lonSec < 10 ? `0${lonSec.toFixed(2)}` : lonSec.toFixed(2);
  const lonStr = `${String(lonDeg).padStart(3, "0")}° ${String(lonMin).padStart(2, "0")}' ${lonSecStr}" ${lonHem}`;

  return [latStr, lonStr];
}

export function generateCorridorPolygon(segments: MTRSegment[]): number[][] {
  if (!segments || segments.length === 0) return [];

  const leftPoints: [number, number][] = [];
  const rightPoints: [number, number][] = [];

  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];
    const lat1 = seg.latitude_dec;
    const lon1 = seg.longitude_dec;
    let lat2 = seg.next_lat_dec;
    let lon2 = seg.next_lon_dec;

    if (lat2 == null || lon2 == null) {
      if (i + 1 < segments.length) {
        lat2 = segments[i + 1].latitude_dec;
        lon2 = segments[i + 1].longitude_dec;
      } else {
        break;
      }
    }

    const widthLeft = seg.width_left_nm != null ? seg.width_left_nm : 5.0;
    const widthRight = seg.width_right_nm != null ? seg.width_right_nm : 5.0;

    const course = calculateBearing(lat1, lon1, lat2, lon2);
    const bearingLeft = (course - 90.0 + 360.0) % 360.0;
    const bearingRight = (course + 90.0) % 360.0;

    const startLeft = destinationPoint(lat1, lon1, widthLeft, bearingLeft);
    const startRight = destinationPoint(lat1, lon1, widthRight, bearingRight);

    const endLeft = destinationPoint(lat2, lon2, widthLeft, bearingLeft);
    const endRight = destinationPoint(lat2, lon2, widthRight, bearingRight);

    leftPoints.push(startLeft);
    if (i === segments.length - 1 || (i + 1 < segments.length && segments[i + 1].next_lat_dec == null)) {
      leftPoints.push(endLeft);
    }

    rightPoints.push(startRight);
    if (i === segments.length - 1 || (i + 1 < segments.length && segments[i + 1].next_lat_dec == null)) {
      rightPoints.push(endRight);
    }
  }

  if (leftPoints.length === 0 || rightPoints.length === 0) return [];

  const polygonCoords: number[][] = [];
  for (const [lat, lon] of leftPoints) {
    polygonCoords.push([lon, lat]);
  }
  for (let j = rightPoints.length - 1; j >= 0; j--) {
    const [lat, lon] = rightPoints[j];
    polygonCoords.push([lon, lat]);
  }

  if (polygonCoords.length > 0) {
    polygonCoords.push(polygonCoords[0]);
  }

  return polygonCoords;
}
